### PR TITLE
#41 test for FkGitHubAuthRedirection

### PR DESCRIPTION
**Pull request by olenagerasimova**
For #41 
- added ctor param `EnvironmentVariables` to obtain env params
- created unit test
- removed puzzle

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#56](https://github.com/g4s8/ghman/issues/56) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#56](https://github.com/g4s8/ghman/issues/56) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CJT4R32Q3/f4176b8b-dc8b-489d-b4d3-6b3211a3e0b1); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                            URLEncoder.encode(
+                                new FormattedText("https://%s/auth", host).asString(),
+                                Charset.forName("UTF-8").name()
+                            )
+                        ).asString()
+                    ),
+                    new StringContains(new FormattedText("client_id=%s", client).asString()),
+                    new StringContains("scope=notifications")
+                )
+            )
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java]: @olenagerasimova this does not give us confidence that it is correctly formatted, I think we should that 1) it is a redirect answer and 2) that the redirect header location contains the correct uri.

File: src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            ).print(),
+            new AllOf<>(
+                new ListOf<>(
+                    new StringContains("https://github.com/login/oauth/authorize"),
+                    new StringContains(
+                        new FormattedText(
+                            "redirect_uri=%s",
+                            URLEncoder.encode(
+                                new FormattedText("https://%s/auth", host).asString(),
+                                Charset.forName("UTF-8").name()
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java]: @olenagerasimova prefer the use of `StandardCharsets.UTF_8`

File: src/main/java/com/g4s8/ghman/web/TkApp.java [hunk 3]
```diff
@@ -56,7 +57,7 @@
                                 "code", ".+",
                                 new TkGitHubAuthorization(data)
                             ),
-                            new FkGitHubAuthRedirection()
+                            new FkGitHubAuthRedirection(new EnvironmentVariables())
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/web/TkApp.java]: @olenagerasimova `TkApp` should be the one taking the `EnvironmentVariables` object and passing it to `FkGitHubAuthRedirection` (and other objects if needed)

**Comment by victornoel** [comment]
@olenagerasimova I've added some comments

File: src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import org.llorllale.cactoos.matchers.Assertion;
+import org.takes.rq.RqFake;
+import org.takes.rs.RsPrint;
+
+/**
+ * Test for {@link FkGitHubAuthRedirection}.
+ * @since 1.0
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+class FkGitHubAuthRedirectionTest {
```
Comment (victornoel) [inline src/test/java/com/g4s8/ghman/web/FkGitHubAuthRedirectionTest.java]: @olenagerasimova class must be final

**Comment by olenagerasimova** [comment]
@victornoel done

**Comment by victornoel** [comment]
@olenagerasimova thanks, but the test is not good enough for me, I would prefer that you actually test the request using something like `RestResponse` or something like that.

Here you are just testing that some string contains some other strings, it'stoo weak as a test, I would like this to properly test that the http code IS a redirection and that there IS a `Location` header with the correct URI, and so on.

**Comment by olenagerasimova** [comment]
@victornoel next time please express yourself more concrete in the beginning of the review for the sake of our time, I've added todo. But basically you are asking me to test `RsRedirect` class, this class belongs to external library and its functionality is supposed to be tested there (according to you the test is [week](https://github.com/yegor256/takes/blob/master/src/test/java/org/takes/rs/RsRedirectTest.java) though).

**Comment by victornoel** [comment]
@olenagerasimova I don't understand what you mean: I am not talking about `RsRedirect`. This is an implementation detail. I want the test for `FkGitHubAuthRedirection` to confirm to me that the behaviour of `FkGitHubAuthRedirection` is the one expected. `FkGitHubAuthRedirection` could be implemented with any framework, but the test is here so that if the implementation changes, the behaviour should not change.

This is not against you @olenagerasimova: the test is currently weak and if you don't know how to make it stronger, it's not a problem, just add a todo (as you did) and everything can go on and the codebase can be stronger.

**Comment by victornoel** [comment]
@olenagerasimova and yes the test of RsRedirect is weak :)

@g4s8 it's good to merge I think

**Comment by olenagerasimova** [comment]
@victornoel I do understand perfectly that it is not agains me and explain what I mean by saying that you are asking me to test `RsRedirect`. 
What is the job of `RsRedirect`? Add http response code 303 and `Location` header. To check this is the job of the external library (`takes`), not ours. Using this class (and library) we assume (or hope), that they (devs of this external library) did their job and class is fully functional. So here, in our code, we are not supposed to (in my opinion) test response code or `Location` header (because they are already tested). 
What we have to check though is that redirection link (created in our code) is correct. In order to that, I think that `RsPrint` approach is ok.

**Comment by victornoel** [comment]
@olenagerasimova in that case your test contains too much things, it should only contain one: that the exact complete string of the redirect uri is correct.

The best would be to:
- extract into a separate object the behaviour of creating the redirect URI and test that exact complete string of the redirect uri is correct
- test `FkGitHubAuthRedirection` only in integration tests

Can you add a todo for this?

**Comment by olenagerasimova** [comment]
@victornoel added todo and simplified current test at least to check that response contains whole redirect link.

**Comment by victornoel** [comment]
@olenagerasimova thx!

@g4s8 it's good to merge

**Comment by 0crat** [comment]
@victornoel[/z](https://www.0crat.com/u/victornoel) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by olenagerasimova** [comment]
@g4s8 ping

**Comment by g4s8** [comment]
@olenagerasimova please check merge conflicts

**Comment by olenagerasimova** [comment]
@g4s8 conflicts resolved in [2c7c91f](https://github.com/olenagerasimova/ghman/commit/2c7c91f9437f115a32def47810d63f954fc972c1), but why nothing changed here?

**Comment by g4s8** [comment]
@olenagerasimova you can try:
```bash
git commit --allow-empty -m "some-message" && git push origin 41
```

**Comment by olenagerasimova** [comment]
@g4s8 thanks, done

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#56](https://github.com/g4s8/ghman/issues/56) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ammaratef45** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)

**Pull request merged**